<!-- inherit template from layout.html-->
{% extends "layout_no_sidebar.html" %}
{% block content %}

<div class="container">
    <meta charset="UTF-8">
    <h3>PhosphoQuest Documentation</h3>
    <hr>
    <H5> Database Contents</H5>
    <br>
    <p>
        Resources in this database were obtained from various sites.
        We thank everyone involved with maintaining, updating and keeping these
        online sites available and free to use for non-commercial purposes.
    </p>
    <p>The majority of the information on this website was obtained from
    <a href="https://www.phosphosite.org">PhosphositePlus</a>. A superb online resource.</p>
    <p>Inhibitors information was obtained from
        <a href="https://www.bindingdb.org"> BindingDB</a>.</p>

    <p>Additional information to fill in gaps in the database was obtained
        via API from <a href="https://www.uniprot.org">Uniprot</a>
        and <a href="https://pubchem.ncbi.nlm.nih.gov/"> PubChem</a>. Particular thanks to
        PubChem for providing the widget resources which are used on the inhibitor detail information
        pages of this website.</p>
    <hr>
    <br>
    <H5> Experimental Data Analysis</H5>
    <br>
    <p>
        What kind of experimental analysis Phosphoquest can handle and why
        users might want to use PhosphoQuest to do their analysis.

    </p>
    <hr>
    <br>
    <H5> Table format for uploading to PhosphoQuest analysis tool.</H5>
    <br>
    <p>
        Please see below for an example of the table format required:
    </p>   
        ### example table image ###
    </p>
        <b>Column 1: </b><em>Substrate</em> column with gene name and 
        site. Note that non-phosphorylated peptides should also be in the 
        format indicated i.e. <em>gene name(None)</em>. However, the analysis 
        tool can also handle substrate entries for non-phosphorylated peptides 
        that are <em>only</em> gene names.
    </p>
        <b>Column 2: </b>Mean intensity of <em>Control</em> replicate intensities.
    </p>
        <b>Column 3: </b>Mean intensity of <em>Treatment/Condition</em> replicate intensities.
    </p>
        <b>Column 4: </b>Fold change calculation: <em>Treatment over the Control.</em> 
        Note: these should not be log2 fold calculations!
    </p>
        <b>Column 5: </b>T-test P-values. Please use <em>uncorrected</em> values,
        as PhosphoQuest will implement correction for multiple-testing errors. 
    </p>
        <b>Column 6-7: </b>Coefficients of variation (CV's) for the <em>Control</em> 
         and <em>Treatment/Condition</em> columns, respectively. PhosphoQuest can also handle
         data that doesn't have these column entries. Please see section -
         <em>Analysis functions</em> below for further details.
    </p> 
    </p> 
    <hr>
    <br>
    <H5> Data analysis - general summary</H5>
    <br>
    <p>
         The following will be an overview of the strategy employed to analyse
         your data. Conceptually, the process can be broken down into 6 steps
         which collate and categorise your data. We then use this analysis to 
         generate data visualisations, we hope will help you interpret your
         phospho-proteomics data.           
    </p>
         <b>Step 1: </b> Data table structure check and basic filtering. This
         pre-analysis step implements a basic error check and determines the format
         of the input table. Data tables without CV columns are processed at this
         stage. Peptide entries with at least one quantitation event are passed 
         for further processing. 
    </p>
         <b>Step 2: </b> A series of sequential steps, categorise the data into
         groupings. For example: which hits are unique or appear in both the 
         <em>Control</em> and <em>Treatment/Condition</em>, which entries are 
         phosphorylated peptides, are cvs within a certain tolerance etc. The
         table is also further filtered for phospho-sites only.
     </p>
         <b>Step 3: </b> P-values corrected for multiple testing errors.
         A default permissable error rate of 0.05 is applied, with only the 
         Benjamini-Hochberg method for multiple testing correction currently implememented.
         Please see this link for background information: 
         https://www.nature.com/articles/nbt1209-1135
    </p>
         <b>Step 4: </b> Analysis in previous steps used to further categorise 
         , filter and sort data. It is at this stage, that the data table is 
         queried against the <em>PhosphoQuest</em> database, determining
         which substrates and sites map to information held on the database.
         Data analysis at this stage allows the visualisation of significantly 
         differentially expressed hits in the form of a styled table and a
         volcano plot.
    </p>
         <b>Step 5: </b> Metrics of the data are extracted for downstream 
         visualisation. For example, phospho pull-down enrichment efficiency 
         & distributions of phosphorylated amino acid residues to name 
         two categories.
    </p>
         <b>Step 6: </b> This analysis step has 2 sub-categories:
         </p><b>A </b>- Determine global freqencies of substrate/sites and their 
         corresponding kinases in data table - this data is visualised using
         wordclouds and frequency charts.
         </p><b>B </b>- Calculate relative kinase activity, in the subset of the data,
         whose hits are considered significantly differentially expressed. This 
         relative activity is determined from the log2 fold changes  
         <em>(Treatment over Control)</em>, calculated for their substrates/sites.
    <hr>
    <br>
    <H5> Output information and  files</H5>
    <br>
    <p>
        Information here about what analysis is produced, and the  output file.
    </p>
</div>
{% endblock content %}