<!-- inherit template from layout.html-->
{% extends "layout_no_sidebar.html" %}
{% block content %}

<div class="container">
    <meta charset="UTF-8">
    <h3>PhosphoQuest Documentation</h3>
    <hr>
    <H5> <b><em>Database Contents</em></b></H5>
    <br>
    <p>
        Resources in this database were obtained from various sites.
        We thank everyone involved with maintaining, updating and keeping these
        online sites available and free to use for non-commercial purposes.
    </p>
    <p>The majority of the information on this website was obtained from
    <a target="_blank" href="https://www.phosphosite.org">PhosphositePlus</a>. A superb online resource.</p>
    <p>Inhibitors information was obtained from
        <a target="_blank" href="http://www.kinase-screen.mrc.ac.uk/kinase-inhibitors"> MRC
            Kinase Profiling Inhibitor Database</a> and
        <a target="_blank" href="https://www.bindingdb.org"> BindingDB</a>.</p>

    <p>Additional information to fill in gaps in the database was obtained
        via API from <a target="_blank" href="https://www.uniprot.org">Uniprot</a>
        and <a target="_blank" href="https://pubchem.ncbi.nlm.nih.gov/"> PubChem</a>. Particular thanks to
        PubChem for providing the widget resources which are used on the inhibitor detail information
        pages of this website.</p>
    <hr>
    <br>
    <H5> <b><em>Background</em></b></H5>
    <br>
    <p>
        PhosphoQuests analytical tool is designed to process data generated
        by phospho-proteomics experiments. Typically, the experimental side 
        would entail tryptic protein digestion, enrichment for phospho-peptides
        using pull down methods employing i.e. titanium dioxide followed by 
        analysis using mass spectrometry.
    <p>
        Your downstream data analysis pipeline, whether it be through packages
        such as MaxQuant, Progenesis QI or in-house tools, should generate 
        identification as well as relative quantitation reports. These outputs
        may take different forms, therefore we provide guidance below on the input 
        format required by the tool. 
    </p>
    <hr>
    <br>
    <H5> <b><em>Pre-requisites for uploading data to PhosphoQuests analysis tool</em></b></H5>
    <br>
    <p>
        Please see below for an example format required by PhosphoQuest:
    </p>   
        <img src="{{url_for('static', filename='example_table_webapp.png')}}"/>
    </p>
        <b>Column 1: </b><em>Substrate</em> column with gene name and 
        site. Note that non-phosphorylated peptides should also be in the 
        format indicated i.e. <em>gene name(None)</em>. However, the analysis 
        tool can also handle substrate entries for non-phosphorylated peptides 
        that are <em>only</em> gene names.
    </p>
        <b>Column 2: </b>Mean intensity of <em>Control</em> replicate intensities.
    </p>
        <b>Column 3: </b>Mean intensity of <em>Treatment/Condition</em> replicate intensities.
    </p>
        <b>Column 4: </b>Fold change calculation: <em>Treatment over the Control.</em> 
        Note: these should not be log2 fold calculations!
    </p>
        <b>Column 5: </b>T-test P-values. Please use <em>uncorrected</em> values,
        as PhosphoQuest will implement correction for multiple-testing errors. 
    </p>
        <b>Column 6-7: </b>Coefficients of variation (CV's) for the <em>Control</em> 
         and <em>Treatment/Condition</em> columns, respectively. PhosphoQuest can also handle
         data that doesn't have these column entries. Please see section:
         <em>Data analysis - general summary</em> below for further details.
    </p> 
    </p> 
    <hr>
    <br>
    <H5> <b><em>Data analysis - general summary</em></b></H5>
    <br>
    <p>
         The following is an overview of the strategy employed to analyse
         your data. Conceptually, the process can be broken down into 6 steps
         which collate and categorise your data. We then use this analysis to 
         generate data visualisations.           
    </p>
         <b>Step 1: </b> Data table structure check and basic filtering. This
         pre-analysis step implements a basic error check and determines the format
         of the input table. Data tables without CV columns are processed at this
         stage. Peptide entries with at least one quantitation event are passed 
         for further processing. 
    </p>
         <b>Step 2: </b> A series of sequential steps, categorise the data into
         groupings. For example: which hits are unique to or appear in both the 
         <em>Control</em> and <em>Treatment/Condition</em>, which entries are 
         phosphorylated peptides, are cvs within a certain tolerance etc. The
         table is also further filtered for phospho-sites only.
     </p>
         <b>Step 3: </b> P-values corrected for multiple testing errors.
         A default permissable error rate of 0.05 is applied, with only the 
         Benjamini-Hochberg method for multiple testing correction currently implememented.
         Please see this link for background information: 
         <a target="_blank" href="https://www.nature.com/articles/nbt1209-1135"> Noble 2009.</a>
    </p>
         <b>Step 4: </b> Analysis implemented in previous steps is used to 
         further categorise, filter and sort data. The data table is at this
         stage queried against the <em>PhosphoQuest</em> database, determining
         which substrates and sites map to information held on the database.
         We can now visualise significantly differentially expressed hits in 
         the form of a styled table and a volcano plot. Please see section:
         <em>Data visualisation</em> below for further details.
    </p>
         <b>Step 5: </b> Metrics of the data are extracted for downstream 
         visualisation. For example, pull-down enrichment efficiency to name
         one such measure.
    </p>
         <b>Step 6: </b> This step has 2 sub-categories:
         </p><b>A </b>- Determine global freqencies of substrate/sites and their 
         corresponding kinases in the data table - this data is visualised using
         wordclouds and frequency charts.
         </p><b>B </b>- Calculate relative kinase activity, in the subset of the data,
         whose hits are considered significantly differentially expressed. This 
         relative activity is determined from the log2 fold changes  
         <em>(Treatment over Control)</em>, calculated for their substrates/sites.
    <hr>
    <br>
    <H5> <b><em>Data visualisation</em></b></H5>
    <br>
    <p>
        We have taken the approach of - <em>more visualisation is a good thing.</em>
        To this end PhosphoQuest generates a number of visuals, that we hope
        will help you interpret and understand your data. These take the
        following form:
    </p>
        <b>Styled table: </b> Data processing steps described earlier, help us
        to pare down your data to a subset that meets the following criteria:
        Significantly differentially expressed with a corrected p-value of 
        <em><=0.05.</em> If CV data is present in the original table, then a 
        secondary threshold is applied i.e. hits with CVs <em><=25%.</em>
    </p>   
         
    </p>    
        
</div>
{% endblock content %}